src/Chemistry.hpp

```diff
@@ -0,0 +1,132 @@
+#ifndef CHEMISTRY_HPP_ // NOLINT
+#define CHEMISTRY_HPP_
+//==============================================================================
+// TwoMomentRad - a radiation transport library for patch-based AMR codes
+// Copyright 2020 Benjamin Wibking.
+// Released under the MIT license. See LICENSE file included in the GitHub repo.
+//==============================================================================
+/// \file Chemistry.hpp
+/// \brief Defines methods for integrating primordial chemical network using Microphysics
+///
+
+#include <array>
+#include <limits>
+
+#include "AMReX.H"
+#include "AMReX_BLassert.H"
+#include "AMReX_Extension.H"
+#include "AMReX_GpuQualifiers.H"
+
+#include "hydro_system.hpp"
+#include "radiation_system.hpp"
+
+#include "burn_type.H"
+#include "eos.H"
+#include "extern_parameters.H"
+
+namespace quokka::chemistry
+{
+template <typename problem_t> void computeChemistry(amrex::MultiFab &mf, const Real dt, const Real max_density_allowed)
+{
+
+	for (amrex::MFIter iter(mf); iter.isValid(); ++iter) {
+		const amrex::Box &indexRange = iter.validbox();
+		auto const &state = mf.array(iter);
+
+		amrex::ParallelFor(indexRange, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+			const Real rho = state(i, j, k, HydroSystem<problem_t>::density_index);
+			const Real xmom = state(i, j, k, HydroSystem<problem_t>::x1Momentum_index);
+			const Real ymom = state(i, j, k, HydroSystem<problem_t>::x2Momentum_index);
+			const Real zmom = state(i, j, k, HydroSystem<problem_t>::x3Momentum_index);
+			const Real Ener = state(i, j, k, HydroSystem<problem_t>::energy_index);
+			const Real Eint = RadSystem<problem_t>::ComputeEintFromEgas(rho, xmom, ymom, zmom, Ener);
+
+			std::array<Real, NumSpec> chem = {-1.0};
+			std::array<Real, NumSpec> inmfracs = {-1.0};
+			Real insum = 0.0_rt;
+
+			for (int nn = 0; nn < NumSpec; ++nn) {
+				chem[nn] = state(i, j, k, HydroSystem<problem_t>::scalar0_index + nn);
+			}
+
+			// do chemistry using microphysics
+
+			burn_t chemstate;
+
+			for (int nn = 0; nn < NumSpec; ++nn) {
+				inmfracs[nn] = chem[nn] * rho / spmasses[nn];
+				chemstate.xn[nn] = inmfracs[nn];
+			}
+
+			// stop the test if we have reached very high densities
+			if (rho > max_density_allowed) {
+				amrex::Abort("Density exceeded max_density_allowed!");
+			}
+
+			// input density and eint in burn state
+			// Microphysics needs specific eint
+			chemstate.rho = rho;
+			chemstate.e = Eint / rho;
+
+			// call the EOS to set initial internal energy e
+			eos(eos_input_re, chemstate);
+
+			// do the actual integration
+			// do it in .cpp so that it is not built at compile time for all tests
+			// which would otherwise slow down compilation due to the large RHS file
+			void chemburner(burn_t chemstate, const Real dt);
```

Comment:
When calling a function in C++, you can't include the return type or the types of the parameters. This is only done when declaring/defining the function.

Reply:
Do you mean changing `void chemburner(burn_t chemstate, const Real dt);` to `void chemburner(chemstate, dt);` when the function `chemburner` is called in `Chemistry.hpp`? If I do that, I get an error (will dig it out). Keeping it the same, the code compiles.

Reply:
Function calls are always like this: `chemburner(chemstate, dt);`

Reply:
I get this error:

```
/g/data1b/jh2/ps3459/quokka/src/Chemistry.hpp:77:25: error: there are no arguments to ‘chemburner’ that depend on a template parameter, so a declaration of ‘chemburner’ must be available [-fpermissive]
   77 |                         chemburner(chemstate, dt);

```

Reply:
Ah, this is because chemburner must be 'declared' before it is 'defined'. You can fix this by copying the function signature `void chemburner(burn_t chemstate, Real dt);` and pasting it near the beginning of Chemistry.hpp.

Reply:
See https://stackoverflow.com/questions/1410563/what-is-the-difference-between-a-definition-and-a-declaration